lib/api/BackbeatAPI.js

```diff
@@ -482,68 +482,59 @@ class BackbeatAPI {
-                        return this._redisClient.zscore(key, member,
-                            (err, score) => {
-                                if (err) {
-                                    return done(err);
-                                }
-                                nextMarker = score;
-                                return done(null, results, index, true);
-                            });
-                    }
-                    return done(null, results, undefined, true);
-                },
-                (results, index, hasLastMember, done) => {
-                    const slice = results.slice(index);
-                    slice.forEach(member =>
-                        entries.push(new ObjectFailureEntry(member, sitename)));
-                    // If the listing result didn't include the last member, we
-                    // should continue checking if there is one in subsequent
-                    // sorted sets to retrieve a next marker value. Otherwise we
-                    // only continue if this isn't the last sorted set to check.
-                    shouldContinue = hasLastMember ?
-                        i < keys.length && entries.length < listingLimit :
-                        i < keys.length;
-                    return done();
-                },
-            ], next);
-        }, () => shouldContinue, err => cb(err, nextMarker, entries));
+            if (score) {
+                // If the request included a next marker we want to list the
+                // sorted set from the member with that score, and we don't want
+                // to use the score in subsequent requests.
+                score = undefined;
+                return this._redisClient
+                    .zrangebyscore(key, '-inf', marker, 'withscores', next);
+            }
+            return this._redisClient.zrange(key, 0, -1, 'withscores', next);
+        }, results => {
+            if (lastMemberScore) {
+                // If the previous sorted set didn't include nextMarker, we are
+                // here only to get the marker.
+                nextMarker = results[results.length - 1];
+                return false;
+            }
+            let resultsIndex;
+            // The results include both the member and its score in sequence.
+            const totalMembers = results.length / 2 + entries.length;
+            if (totalMembers >= listingLimit) {
+                resultsIndex = totalMembers * 2 - listingLimit * 2;
+                lastMemberScore = results[resultsIndex + 1];
+                nextMarker = results[resultsIndex - 1];
+                // We keep searching backward until we find a new score. In this
+                // case, we'll include additional members in the listing. If we
+                // reach the end of this sorted set, nextMarker is undefined and
+                // we go on to the next sorted set to find nextMarker.
+                while (nextMarker === lastMemberScore) {
```

Comment:
I'm thinking this will also need to make sure `resultsIndex` >= 2

Reply:
I'm relying `nextMarker` being `undefined` in that case. Since `lastMemberScore` will be defined, it ends the loop.